File: 栈/71.简化路径.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        # 错误示例

        # stack = []
        # cur = ''
        # for char in path + '/':  # 为了有效划分最后一个
        #     if char == '/':  # 划分每个需要的操作
        #         if cur == '' or cur == '.':
        #             continue # 会跳过第15行代码，当前的cur不会被清空
        #         if cur == '..':
        #             if stack: stack.pop()
        #         else:
        #             stack.append(cur)
        #         cur = ''
        #     else:
        #         cur += char
        #
        # return '/' + '/'.join(stack)
        #
        # 输入
        # path =
        # "/a///b///c/d//././/.."
        # 输出
        # "/a/b"
        # 预期结果
        # "/a/b/c"

        stack = []
        cur = ''
        for char in path + '/':  # 为了有效划分最后一个
            if char == '/':  # 划分每个需要的操作
                if cur == '..':
                    if stack: stack.pop()
                elif cur != '' and cur != '.':
                    stack.append(cur)
                cur = ''
            else:
                cur += char

        return '/' + '/'.join(stack)

File: 栈/71.简化路径.js
/**
 * @param {string} path
 * @return {string}
 */
//* 使用split
var simplifyPath = function (path) {
  const token = path.split("/");
  const stack = [];
  for (const p of token) {
    if (p === "." || p === "") continue;
    else if (p === "..") {
      //! 这里只是一共子情况 如果写 p === ".." && stack.length 和其它情况就不互斥了
      if (stack.length) {
        stack.pop();
      }
    } else {
      stack.push(p);
    }
  }
  return "/" + stack.join("/");
};

//* 不适用split
var simplifyPath = function (path) {
  const stack = [];
  let cur = "";

  path += "/"; // 为了最后一段
  for (const char of path) {
    if (char === "/") {
      // 处理token
      if (cur === "..") {
        if (stack.length) stack.pop();
      } else if (cur !== "." && cur !== "") {
        stack.push(cur);
      }
      cur = "";
    } else {
      cur += char;
    }
  }
  return "/" + stack.join("/");
};
